pkg/simulcastsender.go

```diff
@@ -0,0 +1,214 @@
+package sfu
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/pion/rtp/codecs"
+
+	"github.com/pion/ion-sfu/pkg/log"
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+	"github.com/pion/webrtc/v3"
+)
+
+// WebRTCSimulcastSender represents a Sender which writes RTP to a webrtc track
+type WebRTCSimulcastSender struct {
+	ctx            context.Context
+	cancel         context.CancelFunc
+	onCloseHandler func()
+	sender         *webrtc.RTPSender
+	track          *webrtc.Track
+	rtcpCh         chan rtcp.Packet
+	switchCh       chan uint8
+	maxBitrate     uint64
+	target         uint64
+
+	simulcastSSRC uint32
+	lTS           uint32
+	bTS           uint32
+	lSN           uint16
+	bSN           uint16
+	lSSRC         uint32
+	lTSCalc       time.Time
+
+	once sync.Once
+}
+
+// NewWebRTCSimulcastSender creates a new track sender instance
+func NewWebRTCSimulcastSender(ctx context.Context,
+	track *webrtc.Track,
+	sender *webrtc.RTPSender,
+	simulcastSSRC uint32,
+) Sender {
+	ctx, cancel := context.WithCancel(ctx)
+	s := &WebRTCSimulcastSender{
+		ctx:           ctx,
+		cancel:        cancel,
+		sender:        sender,
+		track:         track,
+		maxBitrate:    routerConfig.MaxBandwidth * 1000,
+		rtcpCh:        make(chan rtcp.Packet, maxSize),
+		switchCh:      make(chan uint8, 1),
+		simulcastSSRC: simulcastSSRC,
+	}
+
+	go s.receiveRTCP()
+	return s
+}
+
+// ReadRTCP read rtp packet
+func (s *WebRTCSimulcastSender) ReadRTCP() chan rtcp.Packet {
+	return s.rtcpCh
+}
+
+// WriteRTP to the track
+func (s *WebRTCSimulcastSender) WriteRTP(pkt *rtp.Packet) {
+	// Simulcast write RTP is sync, so the packet can be safely modified and restored
+	if s.ctx.Err() == nil {
+		// Backup pkt original data
+		origPT := pkt.Header.PayloadType
+		origSSRC := pkt.SSRC
+		origSeq := pkt.SequenceNumber
+		origTS := pkt.Timestamp
+		var td uint32
+		// Check if packet SSRC is different from before
+		// if true, the video source changed
+		if s.lSSRC != pkt.SSRC {
+			relay := false
+			// Wait for a keyframe to sync new source
+			switch s.track.Codec().PayloadType {
+			case webrtc.DefaultPayloadTypeVP8:
+				vp8Packet := codecs.VP8Packet{}
+				if _, err := vp8Packet.Unmarshal(pkt.Payload); err == nil {
+					if vp8Packet.Payload[0]&0x01 == 0 {
+						// Packet is a keyframe
+						relay = true
+					}
+				}
```

Comment:
Can we use the partition head checker?

```suggestion
				check := codecs.VP8PartitionHeadChecker{}
				if ok := check.IsPartitionHead(pkt); ok {
				        // Packet is a keyframe
					relay = true
				}
```

Same exists for other VP9/H264